[JIT] Add static method support for TorchBind

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51177 [JIT] Add static method support for TorchBind**

**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

Differential Revision: [D26356942](https://our.internmc.facebook.com/intern/diff/D26356942)

Commit:
[JIT] Add static method support for TorchBind

**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add static method support for TorchBind"


**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add static method support for TorchBind"


**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add static method support for TorchBind"


**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/python/python_custom_class.cpp:
```diff
@@ -25,11 +25,47 @@ py::object ScriptClass::__call__(py::args args, py::kwargs kwargs) {
+            auto type = self.class_type_.type_->castRaw<ClassType>();
+            TORCH_INTERNAL_ASSERT(type);
+            auto* fn = type->findStaticMethod(name);
+            if (fn) {
```
I think i asked this before, but does the non-static methods live in compilation unit?

SplitInfinity replied:
No, they also live in the TorchBind registry.

SplitInfinity replied:
No, they also live in the TorchBind registry.

tugsbayasgalan commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -765,6 +765,9 @@ std::shared_ptr<SugaredValue> PythonClassValue::attr(
   // Resolve values from the Python object first (e.g. for static methods on
   // this type, resolve them as functions)
+  if (auto* fn = type_->findStaticMethod(field)) {
+    return std::make_shared<FunctionValue>(fn);
```
could you explain why we are doing this?

SplitInfinity replied:
Sure. Let's say we have a TorchBind type `TorchBindClass`. The name `torch.classes.namespace.TorchBindClass` returns an instance of `ScriptClass` with `TorchBindClass`'s JIT type as its `class_type_` member. When compiling a *TorchScript* function that contains `torch.classes.namespace.TorchBindClass`, this is going to be resolved to a regular Python class, so a `PythonClassValue` will be created for it (see line 1121). If `type_->findStaticMethod(field)` is not null (where `type_` is set to `script_class.class_type_` on line 1121), it must be a TorchBind class (for now). In that case, we want to use the static method on the JIT type instead looking one up on the Python object (which we won't be able to compile).

gmagogsfm reviewed (approved):
Looks good. Just a few questions to help with my own understanding.

SplitInfinity commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1405,6 +1405,25 @@ bool ClassType::hasMethod(const std::string& name) const {
+}
+
+torch::jit::Function* ClassType::findStaticMethod(const std::string& name) const {
+  for (auto method : staticmethods_) {
```
When in Rome, do as the Romans do. ;)

I don't think there is measurable benefit to using maps instead but I can file an issue to do this as a follow-up.

SplitInfinity replied:
When in Rome, do as the Romans do ;)

Jokes aside, I can file an issue to fix this (for methods and properties too).

SplitInfinity commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -21,6 +21,9 @@ void inlineCalls(Block* block) {
         auto function_constant = cur->input(0)->node();
         auto fun_type =
             function_constant->output()->type()->expect<FunctionType>();
+        if (!fun_type->function()->isGraphFunction()) {
```
It already existed below for methods:
```
          if (!function.isGraphFunction()) {
            continue;
          }
```

This is the first time we introduced a function (i.e. no `self`) that is not a graph function, so we didn't need it before.

SplitInfinity commented on torch/csrc/jit/api/compilation_unit.h:
```diff
@@ -138,6 +138,18 @@ struct TORCH_API CompilationUnit {
     return ret;
   }
 
+  Function* create_function(
+      c10::QualifiedName name,
+      std::unique_ptr<Function> function,
+      bool shouldMangle = false) {
+    if (shouldMangle) {
+      name = mangle(name);
+    }
+    auto ret = function.get();
+    register_function(std::move(function));
+    return ret;
+  }
+
```
Not used, remove.

Commit:
Update on "[JIT] Add static method support for TorchBind"


**Summary**
This commit adds support for static methods to TorchBind. Just like
pybind, the API for declaring a static method is `def_static(...)`. A
static method must be called on the class directly, and can be called
both in Python as well as TorchScript.

Support for static methods is implemented in a manner similar to that of
instance methods. Registered static functions are wrapped in a layer of
unboxing logic, their schemas are inferred using templates and
metaprogramming, and they are added to the `ClassType` object
corresponding to the TorchBind class on which they are registered.
ScriptClass has been extended to support a `__getattr__` function so
that static methods of TorchBind classes can be invoked in Python. The
implementation of `__getattr__` returns `ScriptClassFunctionPtr`, a
version of `StrongFunctionPtr` without a compilation unit (since the
functions of a TorchBind class live inside the TorchBind registry).
Within TorchScript, TorchBind static functions are desugared in
`PythonClassValue::attr` by looking them up on the class type of the
`PythonClassValue` instance.

**Test Plan**
This commit adds a unit test that tests a simple static method on a
TorchBind class.

[ghstack-poisoned]
